test_plans/misc_classes.asciidoc

```diff
@@ -0,0 +1,119 @@
+
+==== `nd_range`
+The existing tests are modified in the following way:
+
+- Mark all tests for the constructor that takes an offset as deprecated.
+- Mark all tests for `nd_range::get_offset` as deprecated.
+
+==== `item`
+The existing tests are modified in the following way:
+
+- Mark tests for `item::get_offset` as deprecated.
+- The existing test for `item` when `Dimension == 1` is `item_1d.cpp`. This test launches a kernel with the parameter `sycl::item<1> item`. It is modified by adding a test for `operator size_t() const` (which only is available when `Dimension == 1`):
+* Call the above operator on `item` and check that the result is equal to `item.get_id(0)`.
+* Let `item_const` be a `const` copy of `item`. Call the above operator on `item_const` and check that the result is equal to `item.get_id(0)`.
+- Check that `std::is_default_constructible_v<sycl::item<D>>` is false for `D = 1, 2, 3`, since the constructor is specified as `item() = delete;`.
+
+==== `nd_item`
+The existing tests are modified in the following way:
+
+- Remove all tests relating to `nd_item::barrier` and `nd_item::mem_fence`.
+- Mark tests for `nd_item::get_offset` as deprecated.
+- Check that `std::is_default_constructible_v<sycl::nd_item<D>>` is false for `D = 1, 2, 3`, since the constructor is specified as `nd_item() = delete;`.
+
+==== `h_item`
+The existing tests are modified in the following way:
+
+- Check that `std::is_default_constructible_v<sycl::h_item<D>>` is false for `D = 1, 2, 3`, since the constructor is specified as `h_item() = delete;`.
+
+=== Reduction variables
+The below tests are added.
+
+==== `reducer` class
+Check that the following statements evaluate to `false`:
+
+- `std::is_copy_constructible_v<sycl::reducer>`
+- `std::is_move_constructible_v<sycl::reducer>`
+- `std::is_copy_assignable_v<sycl::reducer>`
+- `std::is_move_assignable_v<sycl::reducer>`
+
+Since `sycl::reducer` may have implementation-defined template parameters, use `decltype`.
+
+==== `reducer` member functions
+This test is repeated for the types of the existing reduction tests: `char`, `signed char`, `unsigned char`, `short int`, `unsigned short int`, `int`, `unsigned int`, `long int`, `unsigned long int`, `float`, `long long int`, `unsigned long long int`. In addition, if the device has `aspect::fp64`: `double`. In addition, if the device has `aspect::fp16`: `sycl::half`.
+
+===== Subscript operator
+This test is repeated for a shared memory pointer, a `sycl::buffer`, and a `sycl::span` over a shared memory pointer:
+
+- Instantiate a `sycl::reduction`, `reduction`, over the data with `sycl::plus`.
+- Launch a kernel `parallel_for(range<1>, reduction, [](id<1>, auto& reducer))`.
+- Check that `decltype(reducer)::value_type` is equal to the value type of the test.
+- Check that `decltype(reducer)::binary_operation` is equal to `sycl::plus`.
+- If using a shared memory pointer or a buffer:
+* Check that `decltype(reducer)::dimensions` is `0`.
```

Comment:
This is correct.  For buffers, the dimensionality of the `reducer` class is always 0, because the most common `reduction` pattern is this one (example taken from the spec):

```c++
myQueue.submit([&](handler& cgh) {
  // Input values to reductions are standard accessors
  auto inputValues = valuesBuf.get_access<access_mode::read>(cgh);

  // Create temporary objects describing variables with reduction semantics
  auto sumReduction = reduction(sumBuf, cgh, plus<>());
  auto maxReduction = reduction(maxBuf, cgh, maximum<>());

  // parallel_for performs two reduction operations
  // For each reduction variable, the implementation:
  // - Creates a corresponding reducer
  // - Passes a reference to the reducer to the lambda as a parameter
  cgh.parallel_for(range<1> { 1024 }, sumReduction, maxReduction,
                   [=](id<1> idx, auto& sum, auto& max) {
                     // plus<>() corresponds to += operator, so sum can be
                     // updated via += or combine()
                     sum += inputValues[idx];

                     // maximum<>() has no shorthand operator, so max can only
                     // be updated via combine()
                     max.combine(inputValues[idx]);
                   });
});
```

Note that:
- `sumBuf` and `maxBuf` are buffers containing a single value.
- The dimensionality of the buffers is 1 (because SYCL stores scalars in 1-dimensional buffers).
- The `reducer` is 0-dimensional in order to enable developers to write `sum +=` rather than `sum[0] +=`

The only way to create a 1-dimensional `reducer` is to create a `reduction` associated with a `span`.  Creating a `reducer` with more than 1 dimension isn't currently supported, but should be supported once we have `mdspan`.

Reply:
I created this PR to help clarify the dimensionality: KhronosGroup/SYCL-Docs#345.  Further comments welcome in that PR.